storagenode/pieces/store.go

```diff
@@ -273,6 +274,78 @@ func (store *Store) Delete(ctx context.Context, satellite storj.NodeID, pieceID
 	return Error.Wrap(err)
 }
 
+// DeleteWithStorageFormat deletes the specified piece only at the specified format version
+func (store *Store) DeleteWithStorageFormat(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	err = store.blobs.DeleteWithStorageFormat(ctx, storage.BlobRef{
+		Namespace: satellite.Bytes(),
+		Key:       pieceID.Bytes(),
+	}, formatVer)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// delete records in both the piece_expirations and pieceinfo DBs, wherever we find it.
```

Comment:
We should probably delete from the DBs _first_. Since we have to have situations where one succeeds and the other fails, we have two options:

- have a possibility of a db row existing but the corresponding file not existing on disk
- have a possibility of a file on disk existing but the corresponding db row not existing

Everything else in the system that accounts for mismatches between the two expects the second but not the first, because the second can already happen naturally with canceled uploads and so on.

Reply:
I did it in this order to match the `Delete` function above. Do you recommend I change that one as well? (It seems like, in either case, it makes sense for them to work in a similar manner.)

Reply:
Another good point. That was a good reason to do it this way. But yeah, that one should change too. I'm pretty sure we don't have any considerations or cleanups for rows in the db that don't have corresponding files.

Reply:
I just realized that it doesn't make sense to have a `DeleteWithStorageFormat` here. In the migrate function, I don't actually want to use it, because I _don't_ want to delete a record from the expiration db.

So I've moved the relevant logic into the migrate function, so it removes from the v0 store and deletes the v0 piece.